### PR TITLE
Findbugs: null pointer dereference fixes.

### PR DESCRIPTION
**Pull request by rsandell**
The bump in core version introduced @CheckforNull on Jenkins.getInstance
and escalated to PluginImpl.getInstance.
Added some static helper/shortcut methods in PluginImpl so not too many
new if statements littering the rest of the code.

This commit should fix all current, including old,
"possible null pointer dereference" warnings.

### CODE REVIEW

**Comment by rsandell** [comment]
Hoping to get some review help. Not sure I have all annotations set correctly for example...

@TWestling @jglick @stephenc @recampbell ?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java [hunk 1]
```diff
@@ -138,7 +142,9 @@ public String getDisplayName() {
          * @return the list of descriptors containing GerritServer's descriptor.
          */
         public static DescriptorExtensionList<GerritServer, GerritServer.DescriptorImpl> serverDescriptorList() {
-            return Jenkins.getInstance()
+            Jenkins jenkins = Jenkins.getInstance();
+            assert jenkins != null;
```
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java]: Note that it _is_ valid for `Jenkins.getInstance` to be null during shutdown, so in general you need to think about whether your code might be running in this mode and behave gracefully if so.

There is no good solution for `DescriptorExtensionList`. For `ExtensionList` you can use the new `lookup` method to behave smoothly during shutdown.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java [hunk 2]
```diff
@@ -126,11 +130,20 @@ public Api getApi() {
...
      * @return the instance.
      */
+    @CheckForNull
     public static PluginImpl getInstance() {
-        return Jenkins.getInstance().getPlugin(PluginImpl.class);
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins != null) {
+            return jenkins.getPlugin(PluginImpl.class);
+        } else {
+            logger.error("INITIALIZATION Error, Jenkins could not be found, so no plugin!");
```
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/PluginImpl.java]: Probably better to just shut up.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/dependency/DependencyQueueTaskDispatcher.java [hunk 3]
```diff
@@ -102,8 +101,13 @@ public DependencyQueueTaskDispatcher() {
...
      */
     public static DependencyQueueTaskDispatcher getInstance() {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
+            logger.error("INITIALIZATION ERROR? Could not find the Jenkins instance.");
+            return null;
+        }
         ExtensionList<DependencyQueueTaskDispatcher> dispatchers =
-                Jenkins.getInstance().getExtensionList(DependencyQueueTaskDispatcher.class);
+                jenkins.getExtensionList(DependencyQueueTaskDispatcher.class);
```
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/dependency/DependencyQueueTaskDispatcher.java]: Use `ExtensionList.lookup` if your baseline is sufficiently new.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java [hunk 4]
```diff
@@ -81,10 +86,10 @@ public static ToGerritRunListener getInstance() {
     }
 
     @Override
-    public synchronized void onCompleted(AbstractBuild r, TaskListener listener) {
+    public synchronized void onCompleted(AbstractBuild r, @Nonnull TaskListener listener) {
         GerritCause cause = getCause(r);
         logger.debug("Completed. Build: {} Cause: {}", r, cause);
-        if (cause != null) {
+        if (r != null && cause != null) {
```
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java]: Why?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/GerritTriggerParametersTest.java [hunk 5]
```diff
@@ -76,15 +64,12 @@
...
     public void setup() {
-        mockStatic(PluginImpl.class);
-        PluginImpl plugin = mock(PluginImpl.class);
-        when(PluginImpl.getInstance()).thenReturn(plugin);
-
         config = Setup.createConfig();
         GerritServer server = new GerritServer(PluginImpl.DEFAULT_SERVER_NAME);
         server.setConfig(config);
-        when(plugin.getServer(eq(PluginImpl.DEFAULT_SERVER_NAME))).thenReturn(server);
-        when(plugin.getFirstServer()).thenReturn(server);
```
Comment (jglick) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/GerritTriggerParametersTest.java]: Any particular reason you are deleting all this?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java (hunk 4 above)
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java]: Minimal effort to get rid of the warning that r might be null.
The method in RunListener is declared as ```public void onCompleted(R r, @Nonnull TaskListener listener) {}```

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/GerritTriggerParametersTest.java (hunk 5 above)
Comment (rsandell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/GerritTriggerParametersTest.java]: Yes, at some point in time it had become some strange abomination with mixed JenkinsRule and PowerMock, when the test didn't pass I started with simplifying the test code to understand what was going wrong in the tests and that fixed the issues.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/dependency/DependencyQueueTaskDispatcher.java (hunk 3 above)
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/dependency/DependencyQueueTaskDispatcher.java]: The baseline is 1.565.3 which doesn't seem to have it.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java (hunk 1 above)
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java]: That depends on when ```Jenkins.getInstance()``` becomes null, is it before or after ```Plugin.stop()```?

**Comment by rsandell** [comment]
@reviewbybees

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java (hunk 4 above)
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/ToGerritRunListener.java]: But in fact it should be `@Nonnull Run`. Can fix that later in core, but in the meantime I think you can just add this annotation to your override.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java (hunk 1 above)
Comment (jglick) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/GerritManagement.java]: Not sure offhand.

**Pull request merged**